backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py

```diff
@@ -356,31 +387,52 @@ def delete_target_role_access_policy(
         logger.info(
             'Deleting target role IAM policy...'
         )
-        access_point_name = S3ShareManager.build_access_point_name(share)
+        access_point_name = S3AccessPointShareManager.build_access_point_name(share)
         existing_policy = IAM.get_role_policy(
             target_environment.AwsAccountId,
             share.principalIAMRoleName,
-            "targetDatasetAccessControlPolicy",
+            IAM_ACCESS_POINT_ROLE_POLICY,
         )
+        key_alias = f"alias/{dataset.KmsAlias}"
+        kms_client = KmsClient(dataset.AwsAccountId, dataset.region)
+        kms_key_id = kms_client.get_key_id(key_alias)
         if existing_policy:
-            if dataset.S3BucketName in ",".join(existing_policy["Statement"][0]["Resource"]):
-                target_resources = [
-                    f"arn:aws:s3:::{dataset.S3BucketName}",
-                    f"arn:aws:s3:::{dataset.S3BucketName}/*",
-                    f"arn:aws:s3:{dataset.region}:{dataset.AwsAccountId}:accesspoint/{access_point_name}",
-                    f"arn:aws:s3:{dataset.region}:{dataset.AwsAccountId}:accesspoint/{access_point_name}/*"
-                ]
-                for item in target_resources:
-                    existing_policy["Statement"][0]["Resource"].remove(item)
-                if not existing_policy["Statement"][0]["Resource"]:
-                    IAM.delete_role_policy(target_environment.AwsAccountId, share.principalIAMRoleName, "targetDatasetAccessControlPolicy")
-                else:
-                    IAM.update_role_policy(
-                        target_environment.AwsAccountId,
-                        share.principalIAMRoleName,
-                        "targetDatasetAccessControlPolicy",
-                        json.dumps(existing_policy),
-                    )
+            s3_target_resources = [
+                f"arn:aws:s3:::{dataset.S3BucketName}",
+                f"arn:aws:s3:::{dataset.S3BucketName}/*",
+                f"arn:aws:s3:{dataset.region}:{dataset.AwsAccountId}:accesspoint/{access_point_name}",
+                f"arn:aws:s3:{dataset.region}:{dataset.AwsAccountId}:accesspoint/{access_point_name}/*"
+            ]
+            ShareManagerUtils.remove_resource_from_statement(
+                existing_policy["Statement"][0],
+                s3_target_resources
+            )
+
+            kms_target_resources = [
+                f"arn:aws:kms:{dataset.region}:{dataset.AwsAccountId}:key/{kms_key_id}",
+                f"arn:aws:kms:{dataset.region}:{dataset.AwsAccountId}:key/{kms_key_id}/*"
+            ]
+            ShareManagerUtils.remove_resource_from_statement(
```

Comment:
Can we add some conditional check here - if `len(existing_policy["Statement"]) > 1` or something of the like before trying to remove principals from the KMS Permissions section

If I have a dataset with SSE-S3 encryption that I have approved folder shares for - I would not have this second KMS statement because the KMS Key is managed by AWS and is not a KMS CMK. 

Then when trying to revoke the folder share - the share revoke process would fail here when cleaning up because existing_policy["Statement"][1] gets an `IndexError: list index out of range` (I ran into this during my testing)

The same would be good to add for bucket sharing revokes as well

Reply:
Yes, good catch! I hadn't thought about the use case where KMS is managed by AWS. 

I have addressed all the points here except the last one where we need to modify `add_missing_resources_to_policy_statement ` to not run into the list index out of range. This one will need some tinkering, so I have not addressed it yet.

Reply:
Fixed this in the next revision.